core/src/main/scala/org/apache/spark/FutureAction.scala

```diff
@@ -39,9 +39,9 @@ trait FutureAction[T] extends Future[T] {
   // documentation (with reference to the word "action").
 
   /**
-   * Cancels the execution of this action.
+   * Cancels the execution of this action with an optional reason.
    */
-  def cancel(): Unit
+  def cancel(reason: Option[String]): Unit
```

Comment:
As I also commented on MimaExcludes, it is okay to _add_ a new overload which accepts an optional reason, but we should not _remove_ the old overload.

Instead, I think we should keep the zero-argument `def cancel(): Unit` and provide a default implementation which simply calls `cancel(None)`.

We don't need to override the zero-argument `cancel()` in subclasses; those can remain unchanged.

If we do that then we won't have to add in explicit `cancel(reason = None)` in a bunch of places, which will help to shrink the size of this PR.